src/ConstructiveSolidGeometry/VolumePrimitives/Cone.jl

```diff
@@ -1,207 +1,247 @@
-        end
-    φ = mod(T(φMax) - T(φMin), T(2π)) == 0 ? nothing : T(φMin)..T(φMax)
-    z = zMax == -zMin ? T(zMax) : T(zMin)..T(zMax)
-    Cone( T, r, φ, z)
-end
-Cone(rbotMin, rbotMax, rtopMin, rtopMax, φMin, φMax, zMin, zMax) = Cone(; rbotMin = rbotMin, rbotMax = rbotMax, rtopMin = rtopMin, rtopMax = rtopMax, φMin = φMin, φMax = φMax, zMin = zMin, zMax = zMax)
+T: Type of values, e.g. Float64
+CO: ClosedPrimitive or OpenPrimitive <-> whether surface belongs to it or not
 
-function Cone(rbot::R1, rtop::R2, height::H) where {R1<:Real, R2<:Real, H<:Real}
-    T = float(promote_type(R1, R2, H))
-    Cone( T, (T(rbot), T(rtop)), nothing, T(height)/2)
-end
+* `r::TR`: 
+    * TR = Real -> Solid Tube 
+    * TR = (Real, Real) -> Pipe (r_in = r[1], r_out = r[2])
+    * TR = ((Real,), (Real,)) Solid widening Tube  -> (r_bot = r[1][1], r_top = r[1][2])
+    * TR = ((Real,Real), (Real,Real)) Solid widening Pipe ->\n(r_bot_in = r[1][1], r_bot_out = r[1][2], r_top_in = r[2][1], r_top_out = r[2][2])
+    * TR = (Nothing, (Real,Real)) Cone ->\n(r_bot_in = r_bot_out = 0, r_top_in = r[2][1], r_top_out = r[2][2])
+    * TR = ((Real,Real), Nothing) Cone ->\n(r_bot_in = r[1][1], r_bot_out = r[1][2], r_top_in = r_top_out = 0)
+    * ... (+ elliptical cases -> (a, b))
+    * Not all are implemented yet
 
+* `φ::TP`: 
+    * TP = Nothing <-> Full in φ
+    * ...
+* `zH::T`: half hight/length of the cone
+"""
+@with_kw struct Cone{T,CO,RT,TP} <: AbstractVolumePrimitive{T, CO}
+    r::RT = 1
+    φ::TP = nothing
+    hZ::T = 1
 
-#Constructors for Tubes
-Tube(;rMin = 0, rMax = 1, φMin = 0, φMax = 2π, zMin = -1/2, zMax = 1/2) = Cone(rMin, rMax, rMin, rMax, φMin, φMax, zMin, zMax)
-Tube(rMin, rMax, φMin, φMax, zMin, zMax) = Tube(; rMin = rMin, rMax = rMax, φMin = φMin, φMax = φMax, zMin = zMin, zMax = zMax)
-
-function Tube(r::R, height::H) where {R<:Real, H<:Real}
-    T = float(promote_type(R,H))
-    Cone(T, T(r), nothing, T(height)/2)
+    origin::CartesianPoint{T} = zero(CartesianPoint{T})
+    rotation::SMatrix{3,3,T,9} = one(SMatrix{3, 3, T, 9})
 end
 
-function Tube(rMin::R1, rMax::R2, height::H) where {R1<:Real, R2<:Real, H<:Real}
-    T = float(promote_type(R1,R2,H))
-    Cone(T, rMin == 0 ? T(rMax) : T(rMin)..T(rMax), nothing, T(height)/2)
-end
-
-# for Tubes
-get_r_at_z(c::Cone{T, <:Union{T, AbstractInterval{T}}, <:Any, <:Any}, z::Real) where {T} = c.r
-
-# for Cones
-get_r_at_z(c::Cone{T, Tuple{T,T}, <:Any, <:Any}, z::Real) where {T} = _get_r_at_z(c.r[1], c.r[2], c.z, z)
-
-function get_r_at_z(c::Cone{T, Tuple{I,I}, <:Any, <:Any}, z::Real) where {T, I<:AbstractInterval{T}}
-    r1::T = _get_r_at_z(c.r[1].left, c.r[2].left, c.z, z)
-    r2::T = _get_r_at_z(c.r[1].right, c.r[2].right, c.z, z)
-    r1..r2
-end
+Cone{T,CO,RT,TP}( c::Cone{T,CO,RT,TP}; COT = CO,
+            origin::CartesianPoint{T} = b.origin,
+            rotation::SMatrix{3,3,T,9} = b.rotation) where {T,CO<:Union{ClosedPrimitive, OpenPrimitive},RT,TP} =
```

Comment:
Default should be `c.origin` and `c.rotation`

Reply:
Indeed, sloppy me..